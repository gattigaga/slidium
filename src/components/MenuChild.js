import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import Icon from "components/Icon";

const Container = styled.div`
  width: 100%;
  padding: 8px 16px;
  box-sizing: border-box;
  cursor: ${({ isClickable }) => (isClickable ? "pointer" : "default")};
  position: relative;
  display: flex;
  align-items: center;
  user-select: none;

  &:hover {
    ${({ isClickable }) => isClickable && "background: #eee;"};
  }
`;

const IconContainer = styled.div`
  width: 32px;
  display: flex;
  position: relative;
`;

const StyledIcon = styled(Icon)`
  font-size: 20px;
  color: ${({ isClickable }) => (isClickable ? "#888" : "#bbb")};
  position: absolute;
  top: 50%;
  left: 0%;
  transform: translateY(-50%);
`;

const Label = styled.span`
  font-size: 12px;
  font-family: Roboto;
  color: ${({ isClickable }) => (isClickable ? "black" : "#bbb")};
`;

const MenuChild = ({ label, icon, isClickable, onClick }) => (
  <Container isClickable={isClickable} onClick={() => isClickable && onClick()}>
    <IconContainer>
      {icon && <StyledIcon name={icon} isClickable={isClickable} />}
    </IconContainer>
    <Label isClickable={isClickable}>{label}</Label>
  </Container>
);

MenuChild.propTypes = {
  icon: PropTypes.string,
  label: PropTypes.string,
  isClickable: PropTypes.bool,
  onClick: PropTypes.func
};

MenuChild.defaultProps = {
  label: "Menu",
  isClickable: true,
  onClick: () => {}
};

export default MenuChild;
