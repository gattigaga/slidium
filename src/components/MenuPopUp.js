import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const Container = styled.div`
  position: absolute;
  min-width: 64px;
  top: ${({ position }) => position.y};
  left: ${({ position }) => position.x};
  background: white;
  border: 1px solid #ccc;
  padding: 8px 0px;
  box-shadow: 1px 1px 8px rgba(0, 0, 0, 0.2);
  z-index: 2;
`;

const MenuPopUp = ({ children, position, onClick }) => (
  <Container position={position} onClick={onClick}>
    {children}
  </Container>
);

MenuPopUp.propTypes = {
  children: PropTypes.node.isRequired,
  position: PropTypes.shape({
    x: PropTypes.string,
    y: PropTypes.string
  }),
  onClick: PropTypes.func
};

MenuPopUp.defaultProps = {
  position: {
    x: "0px",
    y: "0px"
  }
};

export default MenuPopUp;
