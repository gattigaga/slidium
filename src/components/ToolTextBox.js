import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import Tooltip from "components/Tooltip";
import Icon from "components/Icon";

const StyledTooltip = styled(Tooltip)``;

const StyledIcon = styled(Icon)`
  color: #888;
  margin: auto;
  font-size: 14px;
`;

const TextBox = styled.input`
  width: 100%;
  height: 100%;
  box-sizing: border-box;
  background: white;
  padding: 0px 4px;
  font-family: Roboto;
  font-size: 11px;
  font-weight: bold;
  border: 0px;
  outline-width: 0px;
  color: #888;

  &:focus {
    border: 1px solid #ddd;
    background: white !important;
  }
`;

const Dropdown = styled.div`
  flex: 0.25;
  height: 100%;
  box-sizing: border-box;
  border-left: 1px solid white;
  background: ${({ isOpen }) => (isOpen ? "#eee" : "white")};
  display: flex;
  cursor: pointer;
  user-select: none;
`;

const Container = styled.div`
  width: 48px;
  height: 32px;
  box-sizing: border-box;
  border-radius: 4px;
  background: white;
  margin: 4px;
  display: flex;
  cursor: pointer;
  user-select: none;
  position: relative;
  display: flex;

  &:hover {
    background: ${({ isOpen }) => (isOpen ? "white" : "#eee")};
  }

  &:hover ${StyledTooltip} {
    opacity: 1;
  }

  &:hover ${Dropdown} {
    background: #eee;
    ${({ isOpen }) => !isOpen && "border-left: 1px solid #ccc;"};
  }

  &:hover ${TextBox} {
    background: ${({ isOpen }) => (isOpen ? "white" : "#eee")};
  }
`;

TextBox.displayName = "TextBox";
Dropdown.displayName = "Dropdown";

class ToolTextBox extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isOpen: false
    };

    this.toggleDropdown = this.toggleDropdown.bind(this);
  }

  /**
   * Toggle dropdown open or close
   *
   * @memberof ToolTextBox
   */
  toggleDropdown() {
    this.setState(prevState => {
      const { isOpen } = prevState;

      if (!isOpen) {
        this.textBox.select();
      }

      return {
        isOpen: !isOpen
      };
    });
  }

  render() {
    const { isOpen } = this.state;
    const { text, tooltip, onChange, children } = this.props;

    return (
      <Container isOpen={isOpen}>
        {!isOpen && tooltip && <StyledTooltip text={tooltip} />}
        <TextBox
          type="text"
          value={text}
          onChange={onChange}
          onClick={this.toggleDropdown}
          innerRef={ref => {
            this.textBox = ref;
          }}
        />
        <Dropdown onClick={this.toggleDropdown} isOpen={isOpen}>
          <StyledIcon name="arrow_drop_down" />
        </Dropdown>
        {isOpen && children}
      </Container>
    );
  }
}

ToolTextBox.propTypes = {
  text: PropTypes.string.isRequired,
  tooltip: PropTypes.string,
  onChange: PropTypes.func,
  children: PropTypes.node.isRequired
};

export default ToolTextBox;
