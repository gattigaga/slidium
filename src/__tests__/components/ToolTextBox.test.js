import React from "react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";

import ToolTextBox from "components/ToolTextBox";
import MenuPopUp from "components/MenuPopUp";
import MenuChild from "components/MenuChild";

let wrapper, onChange;

beforeEach(() => {
  onChange = jest.fn();
  wrapper = shallow(
    <ToolTextBox text="32" onChange={onChange}>
      <MenuPopUp position={{ y: "100%" }}>
        <MenuChild label="24" isWithoutIcon />
        <MenuChild label="32" isWithoutIcon />
        <MenuChild label="48" isWithoutIcon />
      </MenuPopUp>
    </ToolTextBox>
  );
});

describe("ToolTextBox", () => {
  it("renders without tooltip", () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("renders with tooltip", () => {
    wrapper.setProps({
      tooltip: "Font Size"
    });

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("toggles dropdown", () => {
    wrapper = mount(
      <ToolTextBox text="32" onChange={onChange}>
        <MenuPopUp position={{ y: "100%" }}>
          <MenuChild label="24" isWithoutIcon />
          <MenuChild label="32" isWithoutIcon />
          <MenuChild label="48" isWithoutIcon />
        </MenuPopUp>
      </ToolTextBox>
    );

    // Open dropdown
    wrapper.find("Dropdown").simulate("click");
    expect(toJSON(wrapper)).toMatchSnapshot();

    // Close dropdown
    wrapper.find("Dropdown").simulate("click");
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("calls 'onChange'", () => {
    const event = {
      target: {
        value: "64"
      }
    };

    expect(onChange.mock.calls.length).toEqual(0);
    wrapper.find("TextBox").simulate("change", event);
    expect(onChange.mock.calls.length).toEqual(1);
    expect(onChange).toBeCalledWith(event);
  });
});
