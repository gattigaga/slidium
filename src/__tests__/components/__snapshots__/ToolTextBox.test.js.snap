// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolTextBox renders with tooltip 1`] = `
<styled.div
  isOpen={false}
>
  <Styled(Tooltip)
    text="Font Size"
  />
  <TextBox
    innerRef={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    type="text"
    value="32"
  />
  <Dropdown
    isOpen={false}
    onClick={[Function]}
  >
    <Styled(Icon)
      name="arrow_drop_down"
    />
  </Dropdown>
</styled.div>
`;

exports[`ToolTextBox renders without tooltip 1`] = `
<styled.div
  isOpen={false}
>
  <TextBox
    innerRef={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    type="text"
    value="32"
  />
  <Dropdown
    isOpen={false}
    onClick={[Function]}
  >
    <Styled(Icon)
      name="arrow_drop_down"
    />
  </Dropdown>
</styled.div>
`;

exports[`ToolTextBox toggles dropdown 1`] = `
<ToolTextBox
  onChange={[Function]}
  text="32"
>
  <styled.div
    isOpen={true}
  >
    <div
      className="sc-gzVnrw esMmzi"
    >
      <TextBox
        innerRef={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        type="text"
        value="32"
      >
        <input
          className="sc-EHOje dOXpNc"
          onChange={[Function]}
          onClick={[Function]}
          type="text"
          value="32"
        />
      </TextBox>
      <Dropdown
        isOpen={true}
        onClick={[Function]}
      >
        <div
          className="sc-bZQynM dlfBam"
          onClick={[Function]}
        >
          <Styled(Icon)
            name="arrow_drop_down"
          >
            <Icon
              className="sc-ifAKCX cVMWkQ"
              name="arrow_drop_down"
            >
              <i
                className="material-icons sc-ifAKCX cVMWkQ"
              >
                arrow_drop_down
              </i>
            </Icon>
          </Styled(Icon)>
        </div>
      </Dropdown>
      <MenuPopUp
        position={
          Object {
            "y": "100%",
          }
        }
      >
        <styled.div
          position={
            Object {
              "y": "100%",
            }
          }
        >
          <div
            className="sc-htoDjs cFrZYG"
          >
            <MenuChild
              isClickable={true}
              isWithoutIcon={true}
              label="24"
              onClick={[Function]}
            >
              <styled.div
                isClickable={true}
                onClick={[Function]}
              >
                <div
                  className="sc-dnqmqq jfRWuj"
                  onClick={[Function]}
                >
                  <styled.span
                    isClickable={true}
                  >
                    <span
                      className="sc-gqjmRU jZwWJw"
                    >
                      24
                    </span>
                  </styled.span>
                </div>
              </styled.div>
            </MenuChild>
            <MenuChild
              isClickable={true}
              isWithoutIcon={true}
              label="32"
              onClick={[Function]}
            >
              <styled.div
                isClickable={true}
                onClick={[Function]}
              >
                <div
                  className="sc-dnqmqq jfRWuj"
                  onClick={[Function]}
                >
                  <styled.span
                    isClickable={true}
                  >
                    <span
                      className="sc-gqjmRU jZwWJw"
                    >
                      32
                    </span>
                  </styled.span>
                </div>
              </styled.div>
            </MenuChild>
            <MenuChild
              isClickable={true}
              isWithoutIcon={true}
              label="48"
              onClick={[Function]}
            >
              <styled.div
                isClickable={true}
                onClick={[Function]}
              >
                <div
                  className="sc-dnqmqq jfRWuj"
                  onClick={[Function]}
                >
                  <styled.span
                    isClickable={true}
                  >
                    <span
                      className="sc-gqjmRU jZwWJw"
                    >
                      48
                    </span>
                  </styled.span>
                </div>
              </styled.div>
            </MenuChild>
          </div>
        </styled.div>
      </MenuPopUp>
    </div>
  </styled.div>
</ToolTextBox>
`;

exports[`ToolTextBox toggles dropdown 2`] = `
<ToolTextBox
  onChange={[Function]}
  text="32"
>
  <styled.div
    isOpen={false}
  >
    <div
      className="sc-gzVnrw kMVusV"
    >
      <TextBox
        innerRef={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        type="text"
        value="32"
      >
        <input
          className="sc-EHOje dOXpNc"
          onChange={[Function]}
          onClick={[Function]}
          type="text"
          value="32"
        />
      </TextBox>
      <Dropdown
        isOpen={false}
        onClick={[Function]}
      >
        <div
          className="sc-bZQynM bjHajs"
          onClick={[Function]}
        >
          <Styled(Icon)
            name="arrow_drop_down"
          >
            <Icon
              className="sc-ifAKCX cVMWkQ"
              name="arrow_drop_down"
            >
              <i
                className="material-icons sc-ifAKCX cVMWkQ"
              >
                arrow_drop_down
              </i>
            </Icon>
          </Styled(Icon)>
        </div>
      </Dropdown>
    </div>
  </styled.div>
</ToolTextBox>
`;
